daffodil-lib/src/main/scala/edu/illinois/ncsa/daffodil/util/DecimalUtils.scala

```diff
@@ -0,0 +1,223 @@
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * CONTRIBUTORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS WITH THE
+ * SOFTWARE.
+ */
+
+package edu.illinois.ncsa.daffodil.util
+
+import com.ibm.as400.access.AS400PackedDecimal
+
+import java.math.{ BigInteger => JBigInteger, BigDecimal => JBigDecimal }
+
+object DecimalUtils {
+
+  def packedToBigDecimal(packedNum: Array[Byte], scale: Int): JBigDecimal = {
+    val packed: AS400PackedDecimal = new AS400PackedDecimal((packedNum.length * 2) - 1, scale)
+
+    packed.toObject(packedNum).asInstanceOf[JBigDecimal]
+  }
+
+  def packedFromBigDecimal(num: JBigDecimal, nBits: Int): Array[Byte] = {
+    val packed: AS400PackedDecimal = new AS400PackedDecimal(nBits/4 - 1, num.scale)
+
+    packed.toBytes(num)
+  }
+
+  def bcdToBigInteger(bcdNum: Array[Byte]): JBigInteger = {
+    val numDigits: Int = bcdNum.size * 2  // 2 digits stored per byte
+    val outputData: Array[Char] = new Array[Char](numDigits)
+    var outputPos: Int = 0
+    var offset: Int = 0
+    var nibble: Int = 0
+
+    while (offset < bcdNum.size) {
+      nibble = (bcdNum(offset) & 0xFF) >>> 4
+      if (nibble > 0x09)
+        throw new NumberFormatException("Invalid high nibble")
+
+      outputData(outputPos) = (nibble | 0x0030).toChar
+      outputPos = outputPos + 1
+
+      nibble = (bcdNum(offset) & 0x0F)
+      offset = offset + 1
+      if (nibble > 0x09)
+        throw new NumberFormatException("Invalid low nibble")
+
+      outputData(outputPos) = (nibble | 0x0030).toChar
+      outputPos = outputPos + 1
+    }
+
+    new JBigInteger(new String(outputData))
+  }
+
+  def bcdToBigDecimal(bcdNum: Array[Byte], scale: Int): JBigDecimal = {
+    new JBigDecimal(bcdToBigInteger(bcdNum), scale)
+  }
+
+  def bcdFromBigInteger(num: JBigInteger, nBits: Int): Array[Byte] = {
+    val inChars: Array[Char] = num.toString.toCharArray
+    var offset: Int = 0
+    var inPos: Int = 0
+    var firstNibble: Int = 0
+    var secondNibble: Int = 0
+    val outArray: Array[Byte] = new Array[Byte](inChars.length / 2 + 1)
+    val leadingZeros: Int = if(inChars.length % 2 == 0) (nBits/4 - inChars.length + 1) else (nBits/4 - inChars.length)
+
+    // Add leading zeros if necessary
+    while(offset < (leadingZeros - 1)) {
+      outArray(offset) = 0
+      offset = offset + 1
+    }
+
+    // Need even number of digits, pad with 0 if necessary
+    if (inChars.length % 2 != 0) {
+      outArray(offset) = (inChars(inPos) & 0x000F).asInstanceOf[Byte]
+      offset = offset + 1
+      inPos = inPos + 1
+    }
+
+    while (inPos < inChars.length) {
+      firstNibble = (inChars(inPos) & 0x000F) << 4
+      inPos = inPos + 1
+      secondNibble = inChars(inPos) & 0x000F
```

Comment:
firstNibble and secondNibble can be vals defines inside the while loop. Don't need to be vars.

Reply:
Make sure to address this comment. var's should be avoided when possible. They make sense for things that need to increment but most other things can be val's. Might require some renaming (e.g. nibble becomes val highNibble val lowNibble), but is the preferred way to do it. Double check these functions for use of vars that could be vals.